app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheetControllerV2.kt

```diff
@@ -430,31 +429,59 @@ class BloodPressureEntrySheetControllerV2 @Inject constructor(
         .withLatestFrom(screenChanges)
         .filter { (_, screen) -> screen == DATE_ENTRY }
 
+    val newBpDataStream = Observables.combineLatest(dateValidations, bpValidations, patientUuidStream)
+        .map { (dateResult, bpResult, patientUuid) ->
+          when {
+            dateResult is Valid && bpResult is Success -> NewBpData.ReadyToSave(dateResult.parsedDate, bpResult, patientUuid)
+            else -> NewBpData.NeedsCorrection
+          }
+        }
+
+    val updateBpDataStream = Observables.combineLatest(dateValidations, bpValidations, existingBpUuidStream)
+        .map { (dateResult, bpResult, bpUuid) ->
+          when {
+            dateResult is Valid && bpResult is Success -> UpdateBpData.ReadyToSave(dateResult.parsedDate, bpResult, bpUuid)
+            else -> UpdateBpData.NeedsCorrection
+          }
+        }
+
     val saveNewBp = saveClicks
-        .withLatestFrom(validDates, validBps, patientUuidStream) { _, date, bp, patientUuid -> Triple(date, bp, patientUuid) }
+        .withLatestFrom(newBpDataStream) { _, newBp -> newBp }
+        .ofType<NewBpData.ReadyToSave>()
         .flatMapSingle { (date, bp, patientUuid) ->
           val dateAsInstant = date.atStartOfDay(UTC).toInstant()
           bloodPressureRepository.saveMeasurement(patientUuid, bp.systolic, bp.diastolic, dateAsInstant)
         }
         .map { { ui: Ui -> ui.setBpSavedResultAndFinish() } }
 
     val updateExistingBp = saveClicks
-        .withLatestFrom(validDates, validBps, existingBpUuidStream) { _, date, bp, existingBpUuid -> Triple(date, bp, existingBpUuid) }
-        .flatMap { (date, newBp, existingBpUuid) ->
+        .withLatestFrom(updateBpDataStream) { _, updateBp -> updateBp }
+        .ofType<UpdateBpData.ReadyToSave>()
+        .flatMapSingle { (date, updatedBp, existingBpUuid) ->
           bloodPressureRepository.measurement(existingBpUuid)
-              .take(1)
+              .firstOrError()
               .map { existingBp ->
                 val dateAsInstant = date.atStartOfDay(UTC).toInstant()
                 existingBp.copy(
-                    systolic = newBp.systolic,
-                    diastolic = newBp.diastolic,
+                    systolic = updatedBp.systolic,
+                    diastolic = updatedBp.diastolic,
                     createdAt = dateAsInstant,
                     updatedAt = dateAsInstant)
               }
               .flatMapCompletable { bloodPressureRepository.updateMeasurement(it) }
-              .andThen(Observable.just({ ui: Ui -> ui.setBpSavedResultAndFinish() }))
+              .andThen(Single.just({ ui: Ui -> ui.setBpSavedResultAndFinish() }))
         }
 
     return saveNewBp.mergeWith(updateExistingBp)
   }
+
+  sealed class NewBpData {
```

Comment:
What do you think about combining these two sealed classes into one?

```kotlin
sealed class SaveBpData {
    data class ReadyToCreateBp(val parsedDate: LocalDate, val bpResult: Success, val patientUuid: PatientUuid) : SaveBpData()
    data class ReadyToUpdateBp(val parsedDate: LocalDate, val bpResult: Success, val bpUuid: UUID) : SaveBpData()
    object NeedsCorrection : SaveBpData()
}
```

Reply:
Even better. I'll make this change 👍.